accounts/abi/reflect.go

```diff
@@ -126,3 +127,36 @@ func requireUniqueStructFieldNames(args Arguments) error {
 	}
 	return nil
 }
+
+// requireUniqueStructFieldNames makes sure struct field names / tags don't collide.
+func requireUniqueStructFieldNames(typ reflect.Type) error {
+
+	exists := make(map[string]bool)
+
+	for j := 0; j < typ.NumField(); j++ {
+		fieldName := typ.Field(j).Name
+
+		if fieldName[:1] != strings.ToUpper(fieldName[:1]) {
+			// skip private struct fields.
+			continue
+		}
+
+		if tagName, ok := typ.Field(j).Tag.Lookup("abi"); ok {
+			if len(tagName) == 0 {
+				return fmt.Errorf("struct: empty abi tag for field '%s'", fieldName)
+			}
+			if exists[tagName] {
+				return fmt.Errorf("struct: bad abi tag mapping for field '%s'", fieldName)
+			}
+			exists[tagName] = true
+		} else {
+			fieldAsTag := strings.ToLower(fieldName[:1]) + fieldName[1:]
```

Comment:
What does `ToLower` do here? How come you don't just use `fieldName`?

Reply:
Oh, thanks @holiman, `abi:` tags, allows more corner cases that I assumed at the beginning, like allowing events like `event DoIt(uint a, uint A);` or  `event DoIt(uint a, uint);`... I did not realize about it.

I need to be compatible with the previous logic without `abi:` and the new one. I need to work more on this.